### PR TITLE
Don't derive Copy in FFI structs/unions that contain pointers

### PR DESCRIPTION
**Pull request by sdroege**
Clone is still derived to as explicitly cloning the structs is less of a
problem, but implicitly copying them can easily cause double frees.

Fixes https://github.com/gtk-rs/gir/issues/1048

### CODE REVIEW

**Comment by sdroege** [comment]
See https://github.com/gtk-rs/gtk-rs/pull/289 for the regenerate PR.

File: src/analysis/types.rs [hunk 1]
```diff
@@ -233,10 +233,137 @@ pub trait DerivesCopy {
...
+    fn derives_copy(&self, lib: &Library) -> bool {
+        if self.is_incomplete(lib) {
+            return false;
+        }
+
+        for field in self.iter() {
+            if field.is_ptr() {
+                return false;
+            }
+        }
```
Comment (GuillaumeGomez) [inline src/analysis/types.rs:287]:
```suggestion
        if field in self.iter().any(|f| f.is_ptr()) {
            return false;
        }
```

File: src/analysis/types.rs [hunk 2]
```diff
@@ -233,10 +233,137 @@ pub trait DerivesCopy {
...
+    }
+}
+
+impl DerivesCopy for Alias {
+    fn derives_copy(&self, lib: &Library) -> bool {
+        if self.is_incomplete(lib) {
+            return false;
+        }
+
+        !self.is_ptr()
```
Comment (GuillaumeGomez) [inline src/analysis/types.rs:263]:
```suggestion
        !self.is_incomplete(lib) && !self.is_ptr()
```

File: src/analysis/types.rs [hunk 3]
```diff
@@ -233,10 +233,137 @@ pub trait DerivesCopy {
...
+    }
+}
+
+impl DerivesCopy for Field {
+    fn derives_copy(&self, lib: &Library) -> bool {
+        if self.is_incomplete(lib) {
+            return false;
+        }
+
+        !self.is_ptr()
```
Comment (GuillaumeGomez) [inline src/analysis/types.rs:273]:
```suggestion
        !self.is_incomplete(lib) && !self.is_ptr()
```

File: src/analysis/types.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/analysis/types.rs:287]: Actually we can do even better (see below).

File: src/analysis/types.rs [hunk 4]
```diff
@@ -233,10 +233,137 @@ pub trait DerivesCopy {
...
+            return false;
+        }
+
+        for field in self.iter() {
+            if field.is_ptr() {
+                return false;
+            }
+        }
+
+        true
```
Comment (GuillaumeGomez) [inline src/analysis/types.rs:289]:
```suggestion
        !self.is_incomplete(lib) && self.iter().all(|f| !f.is_ptr())
```

File: src/analysis/types.rs [hunk 5]
```diff
@@ -233,10 +233,137 @@ pub trait DerivesCopy {
...
+    }
+}
+
+impl DerivesCopy for Class {
+    fn derives_copy(&self, lib: &Library) -> bool {
+        if self.is_incomplete(lib) {
+            return false;
+        }
+
+        self.fields.as_slice().derives_copy(lib)
```
Comment (GuillaumeGomez) [inline src/analysis/types.rs]:
```suggestion
        !self.is_incomplete(lib) && self.fields.as_slice().derives_copy(lib)
```

File: src/analysis/types.rs [hunk 6]
```diff
@@ -233,10 +233,95 @@ pub trait DerivesCopy {
...
+}
+
+impl DerivesCopy for Class {
+    fn derives_copy(&self, lib: &Library) -> bool {
+        !self.is_incomplete(lib) && self.fields.as_slice().derives_copy(lib)
+    }
+}
+
+impl DerivesCopy for Record {
+    #[allow(clippy::if_same_then_else)]
```
Comment (GuillaumeGomez) [inline src/analysis/types.rs:278]: I don't think it's needed anymore.

**Comment by sdroege** [comment]
Not complete apparently. If a struct contains a field that does not implement `Copy`, we still consider it `Copy` sometimes.

**Comment by GuillaumeGomez** [comment]
Do you have an example by any chance?

**Comment by sdroege** [comment]
```
error[E0204]: the trait `Copy` may not be implemented for this type
    --> glib/sys/src/lib.rs:1044:10
     |
1044 | #[derive(Copy, Clone)]
     |          ^^^^
1045 | pub union GVariantBuilder_u {
1046 |     pub s: GVariantBuilder_u_s,
     |     -------------------------- this field does not implement `Copy`
     |
     = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

Another problem is also:

```
error[E0658]: unions with non-`Copy` fields other than `ManuallyDrop<T>` are unstable
    --> glib/sys/src/lib.rs:1046:5
     |
1046 |     pub s: GVariantBuilder_u_s,
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: see issue #55149 <https://github.com/rust-lang/rust/issues/55149> for more information
```

**Comment by GuillaumeGomez** [comment]
Oh, interesting.

**Comment by sdroege** [comment]
I think considering that limitation for `union`s, we can't actually do anything meaningful here yet...

**Comment by sdroege** [comment]
Closing as blocked on https://github.com/rust-lang/rust/issues/55149

**Pull request closed without merging**